indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.seekablestream;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+import javax.validation.constraints.NotNull;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * class that encapsulates a map of partitionId -> sequenceNumber. Redundant getters
+ * are used for proper Jackson serialization/deserialization when processing terminologies
+ * used by Kafka and kinesis (i.e. topic vs. stream)
+ *
+ * @param <T1> partition id type
+ * @param <T2> sequence number type
+ */
+public class SeekableStreamPartitions<T1, T2>
+{
+  public static final String NO_END_SEQUENCE_NUMBER = "NO_END_SEQUENCE_NUMBER";
+
+  private final String stream;
+  private final Map<T1, T2> map;
+
+  @JsonCreator
+  public SeekableStreamPartitions(
+      @JsonProperty("stream") final String stream,
```

Comment:
not sure am getting this. when am i suppose to provide a stream or topic ? what does it means providing both? 
Can you explain what this class is doing ?

Reply:
I considered using separate `KafkaPartitions` and `KinesisPartitions` initially, but later decided to remove them and just use this generic class instead. The problem is that when deserializing datasource metadata, Kafka uses the field `topic` and Kinesis uses `stream`, so I needed to map both topic and stream to the name of the stream, and that's why both stream and topic are in the constructor, it's meant so that either stream or topic need to be provided.